internal/bbr/filter.go

```diff
@@ -0,0 +1,128 @@
+package bbr
+
+import (
+	"math"
+	"time"
+)
+
+// Size of the bottleneck bandwidth filter's window. The paper suggests
+// 6-10, other than that this is arbitrary.
+const btlBwFilterSize = 6
+
+// Filter that estimates the bottleneck bandwidth.
+type btlBwFilter struct {
+	q        queue[int64]
+	Estimate int64
+}
+
+func newBtlBwFilter() btlBwFilter {
+	return btlBwFilter{
+		q: *newQueue[int64](btlBwFilterSize),
+
+		// We set this to something that is non-zero,
+		// but otherwise won't take precedence over any actual
+		// data we receive:
+		Estimate: 1,
+	}
+}
+
+func (f *btlBwFilter) AddSample(deliveryRate int64) {
+	if f.q.Len() == btlBwFilterSize {
+		f.q.Pop()
+	}
+	f.q.Push(deliveryRate)
+	f.Estimate = f.q.Fold(0, max[int64])
+}
+
+// Filter that estimates the round-trip propagation time.
+type rtPropFilter struct {
+	q          queue[rtPropSample]
+	nextSample rtPropSample
+	Estimate   time.Duration
+}
+
+func newRtPropFilter() rtPropFilter {
+	return rtPropFilter{
+		nextSample: rtPropSample{
+			// Set this to a value that will be removed immediately,
+			// and will be lower priority than any other value, so
+			// it will removed from the queue when a new sample is
+			// added.
+			//
+			// XXX: this still could theoretically do the wrong thing
+			// if somebody's clock is set very wrong. We should find
+			// a better way to do this; maybe ask the user to supply
+			// an initial estimate, before any samples are collected?
+			now: time.Unix(math.MinInt64, 0),
+			rtt: math.MaxInt64,
+		},
+	}
+}
+
+type rtPropSample struct {
+	rtt time.Duration
+	now time.Time
+}
+
+func (f *rtPropFilter) AddSample(sample rtPropSample) {
+	// We want to avoid an un-bounded growing queue for two reasons:
+	//
+	// 1. Space usage
+	// 2. Recomputing the estimate is an O(n) operation, so we want to
+	//    keep n small if we're doing that on each ack.
+	//
+	// We manage this as follows: rather than adding each sample to the queue
```

Comment:
I'm having a bit of trouble parsing this sentence.  Rather than add each sample to the queue, we do what?

Reply:
I made a stab at clarifying it; does the new version make any more sense?

Reply:
Much clearer.  I have just one remaining grammatical nit that makes me backtrack a few times while reading.  What do you think of this?

>We manage this as follows: rather than adding each sample to the queue individually, we compute the minimum RTT for each 1-second window, and add that aggregate value to the queue, which in turn drops samples more than 30 seconds old. This gives the benefits of the sliding window without needing to store each and every sample.

Reply:
LGTM, done.